### ISSUE TITLE
Single-bit images are not decoding properly

### ISSUE THREAD

**Issue by matthew-carroll** [TomRoush/PdfBox-Android#467]
**Describe the bug**
A PDF is filled with images that have single-bit color components. The `Bitmap` that's made available by PdfBox-Android decodes the images to mostly black, with various non-black artifacts.

**PDF example**
Attached
[Fenton.-.Playground.pdf](https://github.com/TomRoush/PdfBox-Android/files/10292681/Fenton.-.Playground.pdf)

**Expected behavior**
![Screen Shot 2022-12-22 at 10 50 55 PM](https://user-images.githubusercontent.com/7259036/209286738-6d8d1ac7-d247-48a8-9c71-81f4001641f1.png)

**Actual behavior**
The page is filled with (mostly) black.

**Environment details:**
 - PdfBox-Android version: [e.g. 2.0.25.0]
 - Android API version: [e.g. API 30]

**Additional context**
If single-bit images aren't supported, then I'd like to implement that on my end. I assume there shouldn't be much involved with choosing black or white based on bit values. However, I'm not sure how to access the image byte data. 

I have a `PDImageXObject` called `image`. When I call `image.createInputStream()`, the resulting input stream's size doesn't seem to match my expectations, given the width, height, and bit depth of the image. Is that stream Flate encoded? If so, how can I decode it?

I tried

```java
ByteArrayOutputStream outStream = new ByteArrayOutputStream();
Filter flateDecoder = FilterFactory.INSTANCE.getFilter(COSName.FLATE_DECODE);
DecodeResult result = flateDecoder.decode(imageInputStream, outStream, new COSDictionary(), 0);
```

But this failed with an error:

```
E/PdfBox-Android( 5865): FlateFilter: stop reading corrupt stream due to a DataFormatException
D/flutter_pdf_renderer( 5865): Error while processing a page: java.io.IOException: java.util.zip.DataFormatException: invalid stored block lengths
E/flutter_pdf_renderer( 5865): Failed to load PDF file as PDDocument: java.util.zip.DataFormatException: invalid stored block lengths
W/System.err( 5865): java.io.IOException: java.util.zip.DataFormatException: invalid stored block lengths
W/System.err( 5865): 	at com.tom_roush.pdfbox.filter.FlateFilter.decode(FlateFilter.java:57)
...
```

**Comment by THausherr** [comment]
`image.createInputStream()` is the decoded stream.

Desktop PDFBox works with that file.

**Comment by matthew-carroll** [comment]
@THausherr can you elaborate on what your response indicates? Are you adding more info to this ticket, or are you saying there's a way to get these images working with PdfBox on Android?

**Comment by THausherr** [comment]
My first line means that you're not using the correct method if you wanted the compressed stream.

My second line means that PDFBox for Android somewhat mirrors the code from PDFBox for desktop. So this means that the desktop code is correct, but that something went wrong when porting it to Android.

It would be interesting to know whether `image.createInputStream()` for /Im31 returns all black (zeroes) or mixed stuff? If it returns all zeroes then the bug is with decoding the stream. If it isn't all black then the bug is with converting the stream to an image object.

**Comment by matthew-carroll** [comment]
@THausherr - Can you tell me what's in that stream? If I call `image.createInputStream()` am I getting raw pixel values? or encoded pixel values (e.g., JPEG)? or a PDF encoded stream?

**Comment by TomRoush** [comment]
I took a look at the PDF in question, and it looks like the image is using the ICCBased colorspace. That isn't supported in PdfBox-Android, and DeviceRGB is used as a fallback. Since the image isn't RGB, you end up with a bit of random color at the top of the image while it's trying to treat the image as RGB and making a mess of it.

Adding ICC support would be a pretty big effort, but I can test and see if it's possible that treating single component ICC images as DeviceGray colorspace would be good enough to show the image. At the very least, an unsupported colorspace should probably fail and not be displayed instead of showing garbage

**Comment by matthew-carroll** [comment]
@TomRoush getting a reasonable fallback here would be good. DeviceGray probably works for this particular case.

In general, we're running into color space issues for many bitmaps, and even for things like fill and stroke colors.

Can you add the ability for developers to at least read the parsed definitions for all the color spaces, such as ICC-Based, Lab, etc? I think in most cases, we could implement the appropriate interpretation of those color spaces on our end, at least for the time being. But right now we're totally blocked by PdfBox-Android internally choosing to ignore the real color space and then replace it with DeviceRGB.

When we access an unsupported color space, we don't seem to get an exception, or any indication that PdfBox-Android is giving us a different color space. We only get a log message to that effect.

So, ideally, I could query the color space and get a data structure with the relevant details. Maybe there's no `toColor(components)` method on that data structure, but at least we'd have access to the information to implement that ourselves.

And we're willing to fund some of the work on this because it's a high-priority blocker for what we're building.

CC @kirkbyo

**Comment by THausherr** [comment]
> @THausherr - Can you tell me what's in that stream? If I call `image.createInputStream()` am I getting raw pixel values? or encoded pixel values (e.g., JPEG)? or a PDF encoded stream?

raw pixel values

**Comment by matthew-carroll** [comment]
@THausherr and those are pixel values encoded at the reported bit depth? i.e., 1 bit per pixel? So when I pull a byte out of that stream, I'm getting 8 pixel values?

**Comment by TomRoush** [comment]
This is fixed with #487. PDICCBased was added and it should handle single bit, RGB, and CMYK images.
Rendered image:
![467_Fenton - Playground pdf-2](https://user-images.githubusercontent.com/4317255/220793978-cc0e3961-55af-4a45-8c16-83c1257ee64b.png)
There's another slight issue with the rendering that isn't fixed yet, the horizontal lines. The PDF is made up of a series of images, top to bottom, and in between each image in the PDF there's 1 row of pixels in the rendering where the alpha is 0. Most likely a rounding issue somewhere, but for now you can get around that by rendering in RGB instead of ARGB
